File: scala/app/src/command/Echo.scala
package command

import resp.Data
import error.CommandExecutionError

extension (c: Command)
  def echo(): Either[CommandExecutionError, Data] = {
    val message = c.arguments.nextOption() match {
      case Some(message) => message
      case None => return Left(CommandExecutionError.WrongArguments)
    }

    // Ensure that no more arguments are provided.
    if(c.arguments.hasNext) {
      return Left(CommandExecutionError.WrongArguments)
    }

    Right(message)
  }

File: typescript/src/command/echo.ts
import { Effect } from "effect"
import { Command } from "./command"
import { WrongArguments, type CommandExecutionError } from "../error"
import type { TData } from "../resp/data"

export class EchoCommand extends Command {
  override handle( ): Effect.Effect<TData, CommandExecutionError> {
    const message = this.args.next( )
    if(message.done)
      return Effect.fail(WrongArguments( ))

    // Ensure that no more arguments are provided.
    if(!this.args.next( ).done)
      return Effect.fail(WrongArguments( ))

    return Effect.succeed(message.value)
  }
}
